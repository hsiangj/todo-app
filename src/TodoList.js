import React, {useState} from 'react';
import NewTodoForm from './NewTodoForm';
import Todo from './Todo';

const TodoList = () => {
  const [todos, setTodos] = useState([]);
  const add = (newTodo) => {
    setTodos(todos => [...todos, newTodo]);
  } 

  const edit = (id, editedTask) => {
    setTodos(todos => todos.map( todo =>
      todo.id === id
      ? {...todo, task: editedTask}
      : todo
    ))
  } 

  const remove = (id) => {
    setTodos(todos => todos.filter(todo => todo.id !== id));
  }

  const todoComponent = todos.map(todo => {
    return (
      <Todo 
      id={todo.id}
      key={todo.id}
      task={todo.task}
      remove={remove}
      edit={edit}
      />
  )})

  return (
    <div>
      <h1>Todo</h1>
      <NewTodoForm addTodo={add}/>
      <ul>{todoComponent}</ul>
    </div>
    
  )
}


export default TodoList;