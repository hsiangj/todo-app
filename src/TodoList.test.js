import { render, fireEvent } from '@testing-library/react';
import TodoList from './TodoList';

it("renders without crashing", function(){
  render(<TodoList />);
});

it("matches snapshot", function(){
  const {asFragment} = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});

function addTodo(todoList, task="test app"){
  // select input field
  const taskInput = todoList.getByLabelText('New task:');
  // add value to input field
  fireEvent.change(taskInput, {target: {value: task}});
  // select 'add' button
  const btn = todoList.getByText('Add');
  // click button to submit form field
  fireEvent.click(btn);
};

it("can add a new task", function(){
  const todoList = render(<TodoList />);
  // no task yet so no edit or remove button yet
  expect(todoList.queryByText('Edit')).not.toBeInTheDocument();
  expect(todoList.queryByText('X')).not.toBeInTheDocument();

  // create a todo
  addTodo(todoList);

  // expect todo to be on the page and form to clear
  expect(todoList.getByText('test app')).toBeInTheDocument();
  expect(todoList.getByText('Edit')).toBeInTheDocument();
  expect(todoList.getByText('X')).toBeInTheDocument();
  expect(todoList.getByLabelText('New task:')).toHaveValue('');
});

it("can edit a task", function(){
  const todoList = render(<TodoList />);
  addTodo(todoList);
  // click edit button
  fireEvent.click(todoList.getByText('Edit'));
  const editInput = todoList.getByDisplayValue('test app');
  fireEvent.change(editInput, {target: {value: 'eat sushi'}});
  // click save button
  fireEvent.click(todoList.getByText('Save'));
  // expect edited task to be present
  expect(todoList.getByText('eat sushi')).toBeInTheDocument();
  expect(todoList.queryByText('test app')).not.toBeInTheDocument();
});

it("can remove a task", function(){
  const todoList = render(<TodoList />);
  addTodo(todoList);
  fireEvent.click(todoList.getByText('X'));
  // expect task to be gone
  expect(todoList.queryByText('test app')).not.toBeInTheDocument();
});
