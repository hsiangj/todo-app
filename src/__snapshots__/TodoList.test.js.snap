// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Todo
    </h1>
    <form>
      <label
        for="todo"
      >
        New task: 
      </label>
      <input
        id="todo"
        name="task"
        type="text"
        value=""
      />
      <button>
        Add
      </button>
    </form>
    <ul />
  </div>
</DocumentFragment>
`;
